Allow pandas.Series as properties values

mstabrin opened the pull request:
# Description
<!-- What does this pull request (PR) do? Why is it necessary? -->
<!-- Tell us about your new feature, improvement, or fix! -->
<!-- If your change includes user interface changes, please add an image, or an animation "An image is worth a thousand words!" -->
<!-- You can use https://www.cockos.com/licecap/ or similar to create animations -->
I tried to use values from a pandas Series as properties for a points layer, but the ordering got messed up during the creation of the Points.

```
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': [1, 2, 3, 4], 'val': [1, 2, 3, 4], 'grp': [2, 1, 1, 2]})

In [3]: for idx, sdf in df.groupby('grp'):
   ...:     viewer.add_points(sdf[['x', 'y']], size=0.1, properties={'val': sdf['val']})
   ...:     

In [4]: viewer.layers[0].properties
Out[4]: {'val': array([nan,  2.])}

In [5]: viewer.layers[1].properties
Out[5]: {'val': array([ 1., nan])}
```

I expected the correct values to be assigned to the val property:

```
In [1]: import pandas as pd

In [2]: df = pd.DataFrame({'x': [1, 2, 3, 4], 'y': [1, 2, 3, 4], 'val': [1, 2, 3, 4], 'grp': [2, 1, 1, 2]})

In [3]: for idx, sdf in df.groupby('grp'):
   ...:     viewer.add_points(sdf[['x', 'y']], size=0.1, properties={'val': sdf['val']})
   ...:     

In [4]: viewer.layers[0].properties
Out[4]: {'val': array([2, 3])}

In [5]: viewer.layers[1].properties
Out[5]: {'val': array([1, 4])} 
```

Therefore, I added the functionality to deal with pandas Series objects in addition to numpy arrays and pandas DataFrames.

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [X] the test suite for my feature covers inputs of DataFrames, dictionary of Series, Dictionary of lists, Dictionary of numpy arrays
- [X] all tests pass with my change

## Final checklist:
- [X] My PR is the minimum possible work for the desired functionality
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/developers/translations.html).

Does the documentation need an adjustment for this type of change or is a pandas Series in your eyes equivalent to an array since lists are also not mentioned?
Current documentation:
```
properties (dict {str: array (N,)}, DataFrame) – Properties for each point. Each property should be an array of length N, where N is the number of points.
```

mstabrin committed:
Allow pandas.Series as properties values

mstabrin committed:
Merge branch 'main' into properties_series_needs_drop_index

mstabrin committed:
Add numpy array to the tests and mixed cases

mstabrin committed:
Merge branch 'properties_series_needs_drop_index' of github.com:mstabrin/napari into properties_series_needs_drop_index

mstabrin committed:
Add series property tests to points layer

andy-sweet commented on napari/layers/utils/layer_utils.py:
```diff
@@ -907,6 +909,11 @@ def _validate_features(
     if isinstance(features, pd.DataFrame):
         features = features.reset_index(drop=True)
+    elif isinstance(features, dict):
+        # One needs to reset pd.Series object's indices as well
+        for key, value in features.copy().items():
+            if isinstance(value, pd.Series):
+                features[key] = value.reset_index(drop=True)
```
```suggestion
        features = {key: np.array(value, copy=False)}
```

This should handle all array likes and things that implement `__array__`, which includes `Series`. I think `copy=False` is fine here as we don't make guarantees about copies and this should avoid an unnecessary copy when a `np.ndarray` was passed.

andy-sweet commented on napari/layers/utils/_tests/test_layer_utils.py:
```diff
@@ -461,3 +462,43 @@ def test_feature_table_from_layer_with_custom_index_and_num_data():
+)
+def test_validate_features_returns_correct_df(params):
+    input_df, output_df = params
+    assert _validate_features(input_df).equals(output_df)
```
Also, I think `_FeatureTable.from_layer` may be a better test target here than `_validate_features` as it is more likely to survive expected deprecations/changes. No way you could know that, but would be good to change.

andy-sweet commented:
> Does the documentation need an adjustment for this type of change or is a pandas Series in your eyes equivalent to an array since lists are also not mentioned?

There will hopefully be quite a few documentation updates soon, though I've been saying that for a while. Depending on the fix in this PR, you should feel free to update the docstrings for `features` and `properties`, but should also feel free not to, as there are a lot.

mstabrin committed:
Adjust according to reviews

mstabrin committed:
Remove debug prints

mstabrin commented:
Thank you for your great input @andy-sweet and @brisvag!
I like the general np.array approach. I did not want to change too much in the beginning so I kept the fix related to my related issue.

- Use np.array to create array_like objects
- I removed the integration test from the `points` layer and the `from_layer` function is now used rather then the `_validate_features` function. Since the `from_layer` function already has a bunch of tests, I only added my failing test case.
- Since properties will be deprecated, I added the test for the properties and for the features keyword.
- I added a comment linking to this pull request

andy-sweet reviewed (approved):
Looks great, I'll merge this in 48 hours unless there are any other comments.

mstabrin commented:
@andy-sweet Thank you :)
